sway: add module

alexarice opened the pull request:
Creates a sway module.
This is mainly copied from the i3 module but has some changes to make it work with sway as well as make the defaults closer the the sway default config file.

~~This is wip because I'd like to make options for `output` and `input` sections of the config~~

teto commented:
wouldn't it be better to use the same generator for both instead ? just adapt what needs to be

alexarice commented:
Possibly though there are quite a few subtle changes everywhere and a few descriptions need i3 replaced by sway. I also tried to make the defaults closer to the sway defaults, which changed some stuff, in particular it has added vim movement bindings everywhere.

Lastly I believe the projects could diverge in the future and therefore it is not necessarily true that changes to the i3 module should affect the sway module.

That being said I could try and have a look at what common ground they do have. Would the best solution be to set the sway options equal to the i3 options and just change what needs to be changed?

alexarice commented:
Further, do you think this module is in the correct place? I have put it under programs as that is where it lies on nixos

kalbasit commented:
> Further, do you think this module is in the correct place? I have put it under programs as that is where it lies on nixos

Why not with the rest of window managers? https://github.com/rycee/home-manager/tree/master/modules/services/window-managers

alexarice commented:
I was copying nixpkgs placement. Happy to be overruled though

alexarice commented:
I have added the input and output modules. This is ready for review now

alexarice commented:
I've moved as much as I think I can into common files. The only thing I have changed for i3 is the gaps module and all that has changed is adding some more options which are supported by i3 gaps

alexarice commented on modules/programs/sway.nix:
```diff
@@ -0,0 +1,954 @@
+        description = "Home row direction key for moving right.";
+      };
+
+      terminal = mkOption {
```
if you use a package, how do you specify the executable name?

uvNikita replied:
@alexarice I think you might be right in both cases. Let's leave it as a string for now.

alexarice commented:
any thoughts on the changes?

uvNikita commented:
Sorry for the delay, will test it and review as soon as I'll get some time!

Can suggest the following folder structure for now. What do you think?

```
services
└── window-managers
    └── i3-sway
        ├── lib
        │   ├── functions.nix
        │   └── options.nix
        ├── i3.nix
        └── sway.nix
```

alexarice commented:
Looks good, I'll change that and rebase onto master

tadeokondrak commented on modules/services/window-managers/i3-sway/sway.nix:
```diff
@@ -0,0 +1,304 @@
+      '';
+    };
+
+    systemdIntegration = mkEnableOption "systemdIntegration";
```
Should this be off by default? I think it'd make more sense enabled by default so other modules can take advantage of it, like a future `mako` module for example.

alexarice replied:
To be honest I'm not sure this is needed and things like `mako` should likely be linked against some other target as they work on things other than `sway`

tadeokondrak replied:
The `sway-session.target` is linked to `graphical-session.target` which is what other services like `redshift` use, and works with the sway systemd integration.

alexarice replied:
My point was that if in the future a `mako` service was made then shouldn't it have `graphical-session` as it's target?

tadeokondrak replied:
Yes, that's what I mean, but it would still require `systemdIntegration` to be enabled manually in sway.

alexarice replied:
oh, will `graphical-session` never start without this? Sorry I've only just started using systemd and I don't have a full grasp on what is going on yet

tadeokondrak replied:
I believe so, for `xsession` it's started here: https://github.com/rycee/home-manager/blob/0dfa1eef25a8c1090ff1e38f7a73123626605da0/modules/xsession.nix#L160

alexarice replied:
Sure, can change this

alexarice replied:
I have changed the default to be `pkgs.stdenv.isLinux`

alexarice replied:
Should this also enable the systemd module or is this not necessary?

rycee commented on modules/services/window-managers/i3-sway/sway.nix:
```diff
@@ -0,0 +1,311 @@
+    home.packages = [ cfg.package ] ++ lib.optional cfg.xwayland pkgs.xwayland;
+    home.file.".config/sway/config".text = configText;
+    systemd.user.targets.sway-session = mkIf cfg.systemdIntegration {
+      Unit = {
```
Shouldn't there be a
```
Requires = [ "graphical-session-pre.target" ];
```
to ensure the pre target kicks off?

Also, do you know why the `Wants` and `After` fields are needed? Should they be added to hm-graphical-session?

alexarice replied:
no idea, I just copied https://github.com/swaywm/sway/wiki/Systemd-integration

alexarice replied:
Would you like me to add that?

uvNikita commented on modules/services/window-managers/i3-sway/sway.nix:
```diff
@@ -0,0 +1,311 @@
+
+      menu = mkOption {
+        type = types.str;
+        default = "${pkgs.dmenu}/bin/dmenu_path | ${pkgs.dmenu}/bin/dmenu | xargs swaymsg exec --";
```
`xargs` can also be replaced with `${pkgs.findutils}/bin/xargs`

uvNikita commented on modules/services/window-managers/i3-sway/i3.nix:
```diff
@@ -0,0 +1,228 @@
+          }
+        '';
+      };
+
```
Too many newlines

uvNikita commented on modules/services/window-managers/i3-sway/sway.nix:
```diff
@@ -0,0 +1,311 @@
+        example = "alacritty";
+      };
+
+      menu = mkOption {
```
Seems like these options could be added to i3 module as well?

alexarice replied:
Does the default i3 config use this construction? Would be happy to move it anyway, I don't think it will break anyone's config

uvNikita replied:
We could set it to `exec ${pkgs.dmenu}/bin/dmenu_run` for i3? And I was also thinking that it might make sense to move `left`, `down`, `right`, `up` and `terminal` options there if we want to declare them.

alexarice commented:
Any update? I think I attempted to answer all the queries

uvNikita commented on modules/services/window-managers/i3-sway/sway.nix:
```diff
@@ -0,0 +1,311 @@
+in
+
+{
+  options.services.sway = {
```
I think we should have something similar to `xsession.windowManager.i3`, maybe `wayland.windowManager.sway`?

alexarice replied:
I believe using `left`, `down`, `right`, `up` would conflict with i3's default bindings. i3 default has `h` bound to split horizontal as sway's default is to use vim keybindings for movement.

alexarice replied:
Terminal and menu could be done though

alexarice replied:
I'll change the option name

alexarice commented:
I couldn't get the user systemd service to work myself.

alexarice commented:
I have changed the package option to accept null so that one can use the `programs.sway.enable` option in nixos to install sway

notgne2 commented:
From how I've used this on Sway it's been perfect, but on `i3` it appends `swaybar_command` to the config and triggers an error.

alexarice commented:
Thanks for letting me know /trying it out. I'll fix this and rebase as soon as I have some wifi

alexarice commented:
I think I have fixed the issue though will have to rebase later as the merge conflict is non trivial

teto commented:
mind rebasing ? I will give it a spin

alexarice commented:
Sorry, have had very little time, I'll try and get it done asap

alexarice commented:
Ok I have rebased

alexarice commented:
Have rebased again, noticed this time that there is a test for i3, would it be worth making one for this?

alexarice commented:
Have removed the last change which added bound `${modifer}+0` to `workspace number 10` as this messed up workspace numbering in sway.

teto commented:
I wanted to try the module but didn't manage to. Maybe you could add some checks/warnings for prequisites. For instance when enabling `services.xserver.displayManager.gdm.nvidiaWayland = true;` at the system level, nixos-rebuild warned me with `hardware.nvidia.modesetting.enable = true;`.

I also tried disabling `xsession = { enable = true; }` that started i3. Any advice/comment on how to try it out ? cheers

alexarice commented:
I personally also use the nixos module for sway, as this links sway with dbus. I am only really using this to generate the config file.

That said, I have the following configuration in my `configuration.nix` to start sway

```
programs.fish = {
      enable = true;
      loginShellInit = ''
        not set -q DISPLAY && test (tty) = /dev/tty1
        and exec env _JAVA_AWT_WM_NONREPARENTING=1 sway
      '';
    };
```

I believe it is recommended to start sway from a tty and not with a display manager. If you just wanted to try it out you could manually switch to a new tty and run `sway` from the terminal

teto commented:
thanks for the tip. I tried with the nvidia drivers and it failed. Seems like nouveau doesn't support the GTX970 so I can't test it even though I would like to merge it :(

alexarice commented:
I'm surprised it didn't work at all, I'm sure I got it to run on a GTX 960 at some point (though it was a bit choppy from memory)

uvNikita commented:
There is an example service file in nixos.wiki for starting sway as a systemd service: https://nixos.wiki/wiki/Sway. Do you think this is something that can help the integration? I might try sway soon too to see how it works to use this module.

alexarice commented:
I hadn't seen that, I could give it a try next time I have some free time

alexarice commented:
My understanding from the sway wiki is that setting up the target and setting up the sway service are two different ways of achieving the same thing. They have a bit there on using a login manager but I don't use one so I'm not really sure how that works. I'm going to change `graphical-session.target` to `graphical-session-pre.target` as I think I typod it originally. Also I'm going to rebase

mweinelt commented:
Awesome, works great. Two nits

- Can we validate the sway config using `sway -C`?
- Can we reload sway using `swaymsg -S /run/user/$UID/sway-ipc.$UID.*.sock reload`?

alexarice commented:
> 
> 
> Awesome, works great. Two nits
> 
>     * Can we validate the sway config using `sway -C`?
> 
>     * Can we reload sway using `swaymsg -S /run/user/$UID/sway-ipc.$UID.*.sock reload`?

Sound like good ideas, do you know how to implement them? I am not sure enough on home-manager internals to know off the top of my head how to either of them but can look when I have some time

mweinelt commented:
Basically some Shell script like here https://github.com/rycee/home-manager/pull/829/files#diff-23b1faca336c6b0b297a1da00fb658acR211-R217. I can whip something up if you like.

alexarice commented:
> 
> 
> Basically some Shell script like here https://github.com/rycee/home-manager/pull/829/files#diff-23b1faca336c6b0b297a1da00fb658acR211-R217. I can whip something up if you like.

Sure

mweinelt commented:
Reloading works with the commit below, you can just squash it into yours. Config validation is weird because even if it finds errors it exits with code 0. Probably worth double-checking and then creating an upstream issue.

```
From 9e1dd63d03215ef7fa49d1840419c49674b5b28d Mon Sep 17 00:00:00 2001
From: Martin Weinelt <hexa@darmstadt.ccc.de>
Date: Tue, 4 Feb 2020 22:50:35 +0100
Subject: [PATCH] sway: add reload when config changes

---
 modules/services/window-managers/i3-sway/sway.nix | 13 +++++++++++--
 1 file changed, 11 insertions(+), 2 deletions(-)

diff --git a/modules/services/window-managers/i3-sway/sway.nix b/modules/services/window-managers/i3-sway/sway.nix
index 85c50bc..1d32b14 100644
--- a/modules/services/window-managers/i3-sway/sway.nix
+++ b/modules/services/window-managers/i3-sway/sway.nix
@@ -205,7 +205,7 @@ let
     }
   '';
 
-  configText = ((if cfg.config != null then with cfg.config; ''
+  configFile = pkgs.writeText "sway.conf" ((if cfg.config != null then with cfg.config; ''
     font pango:${concatStringsSep ", " fonts}
     floating_modifier ${floating.modifier}
     default_border ${if window.titlebar then "normal" else "pixel"} ${toString window.border}
@@ -286,7 +286,16 @@ in
 
   config = mkIf cfg.enable {
     home.packages = optional (cfg.package != null) cfg.package ++ optional cfg.xwayland pkgs.xwayland;
-    home.file.".config/sway/config".text = configText;
+    xdg.configFile."sway/config" = {
+      source = configFile;
+      onChange = ''
+        swaySocket=''${XDG_RUNTIME_DIR:-/run/user/$UID}/sway-ipc.$UID.*.sock
+        if [ -S $swaySocket ]; then
+          echo "Reloading sway"
+          $DRY_RUN_CMD ${cfg.package}/bin/swaymsg -s $swaySocket reload
+        fi
+      '';
+    };
     systemd.user.targets.sway-session = mkIf cfg.systemdIntegration {
       Unit = {
         Description = "sway compositor session";
-- 
2.25.0
```

alexarice commented:
```
error: cannot coerce null to a string, at /home/alex/home-manager/modules/services/window-managers/i3-sway/sway.nix:291:18
```
I get this error with the patch, this is because I've set `cfg.package` to null. I'll put in a hacky fix for now though I'm not sure it is ideal

mweinelt commented on modules/services/window-managers/i3-sway/sway.nix:
```diff
@@ -0,0 +1,311 @@
+    enable = mkEnableOption "sway wayland compositor";
+
+    package = mkOption {
+      type = with types; nullOr package;
```
Not sure why null would be a good fit here. What's the idea?

alexarice replied:
I've installed sway using the nixos option which I believe configures some extra stuff, I didn't want this to be overwritten by the home-manager module

alexarice replied:
This might no longer be necessary, I thought that the nixos sway module messed with some system thing that home-manager couldn't touch but maybe everything can be ported here

mweinelt commented on modules/services/window-managers/i3-sway/sway.nix:
```diff
@@ -0,0 +1,311 @@
+      default = pkgs.sway;
+      defaultText = literalExample "${pkgs.sway}";
+      description = ''
+        Sway package to use. Set to null to not install a package (for example if using the nixos option to install sway)
```
I think that unnecessarily overcomplicates the matter. I believe people should just pick the package where the module they're using resides, not mix and match between home-manager and nixos.

alexarice replied:
I've looked over this and it should be an easy change to move all the nixos features here

alexarice commented:
The nixos module has 
```
# To make a Sway session available if a display manager like SDDM is enabled:
    services.xserver.displayManager.sessionPackages = [ swayPackage ];
```
is there an equivalent for home manager?

mweinelt commented:
I think `home.packages = [ swayPackage ];` is what you're looking for.

alexarice commented:
I thought `home.packages` just installed the package, didn't realise it would also add it to display managers

mweinelt commented:
Oh, that's what you mean. Sorry, I missed the comment and expected it to add a package to the sessions environment.

I don't know if it makes sense for home-manager to be involved with system-wide services like the display-manager though. Does the sway package even ship a desktop file? They don't seem to think that's "the way": https://github.com/swaywm/sway/wiki#login-managers.

alexarice commented:
I agree, I do not use a display manager though I think quite a few people prefer them and if there is an easy way to add the functionality then it is probably worth it?

alexarice commented:
I ported everything over but now get a bug with gpg not finding pinentry which goes away if I use the nixos module, going to try to figure this out later

mweinelt commented:
Possibly related to https://github.com/NixOS/nixpkgs/commit/3d832dee59ed0338db4afb83b4c481a062163771?

alexarice commented:
Very possibly, though it is strange that the nixos sway module somehow fixes this

alexarice commented:
Adding `gnupg.agent.pinentryFlavor = "gnome3"` fixed my issue

alexarice commented:
https://github.com/NixOS/nixpkgs/blob/42a2668eb07b169264bca841a1ae31a11452b178/nixos/modules/programs/gnupg.nix#L11-L20
This code was fixing the issue, not sure how to work round this

mweinelt commented:
https://github.com/rycee/home-manager/blob/master/modules/services/gpg-agent.nix#L133

mweinelt commented:
The CI feedback is stuck, but the CI actually ran through without indicating an issue.

What else needs to be adressed to get this merged?

lovesegfault commented:
Anything I can help with to get this merged @alexarice?

alexarice commented:
> Anything I can help with to get this merged @alexarice?

As far as I am concerned, this is ready to merge

mweinelt commented:
I'm actually using this module on my laptop and it works well. I think it's ready to be merged as well.

lovesegfault commented:
Pinging @rycee

rycee commented on modules/modules.nix:
```diff
@@ -152,7 +152,8 @@ let
     (loadModule ./services/window-managers/bspwm/default.nix { condition = hostPlatform.isLinux; })
-    (loadModule ./services/window-managers/i3.nix { })
+    (loadModule ./services/window-managers/i3-sway/i3.nix { })
+    (loadModule ./services/window-managers/i3-sway/sway.nix { })
```
Could do
```suggestion
    (loadModule ./services/window-managers/i3-sway/sway.nix { condition = hostPlatform.isLinux; })
```
to restrict the module to Linux. Unfortunately the i3 module needs to stay as is for backwards compatibility reasons.

alexarice commented:
anything else that needs changing?

rycee commented:
I'm happy to merge it if @uvNikita is 🙂

uvNikita commented:
I see that home-manager sets some session variables and does `systemctl --user import-environment` in files like `.xsession` and `.xprofile`, but this will not happen when running sway, right?

That said, considering that the module doesn't break existing configurations, I don't mind merging it and improve it in the future.

alexarice commented:
> 
> 
> I see that home-manager sets some session variables and does `systemctl --user import-environment` in files like `.xsession` and `.xprofile`, but this will not happen when running sway, right?

This runs `systemctl --user import-environment` in the sway config and sway shouldn't load `.xsession` or `.xprofile`. I'm not quite sure if that is what you were asking

uvNikita commented:
Sorry, missed the `systemctl --user import-environment` part. In `.xsession` file home-manager also adds some session variables to that command, not sure if it matters here: https://github.com/rycee/home-manager/blob/115e76ae12a81bce5dcd19714cdeaaa8d5ca3ce8/modules/xsession.nix#L78-L86
The resulting line looks something like this:
```
systemctl --user import-environment DBUS_SESSION_BUS_ADDRESS DISPLAY SSH_AUTH_SOCK XAUTHORITY XDG_DATA_DIRS XDG_RUNTIME_DIR XDG_SESSION_ID QT_QPA_PLATFORMTHEME
```
As for `.xprofile`, among other things there is this line: 
```
. "/home/user/.nix-profile/etc/profile.d/hm-session-vars.sh"
```
which sets session variables like `EDITOR`, `XDG_CACHE_HOME` and others. We should probably add similar line to the sway starting script too?

alexarice commented:
I'm not sure, I've been using this for a while now and I definitely know my `EDITOR` variable is set correctly.

rycee commented:
@alexarice Are you starting sway from the login console? If so then the shell would already have loaded the environment variables, including the ones in `hm-session-vars.sh`.

I think this should be OK to merge for now and then we can progressively improve it in the future.

Thumbs up or down and we'll merge tomorrow if the ayes have it 😉

alexarice commented:
I am starting it from the login console

rycee commented:
Great job! Rebased to master in 02d60400035028d4c8affab24557344ca371b9fd.
